xgfield/fieldsky.py

```diff
@@ -0,0 +1,105 @@
+def _get_websky_displacements():
+    # Websky displacement fields
+    try:
+        path2disp = os.environ['LPT_DISPLACEMENTS_PATH']
+    except:
+        path2disp = '/Users/shamik/Documents/Work/websky_datacube/'
+
+    backend.print2log(log, f"Path to displacement files set to {path2disp}", level='usky_info')
+
+    if grid_nside == 768:
+        sxfile = path2disp+'sx1_7700Mpc_n6144_nb30_nt16_no768'
+        syfile = path2disp+'sy1_7700Mpc_n6144_nb30_nt16_no768'
+        szfile = path2disp+'sz1_7700Mpc_n6144_nb30_nt16_no768'
+    else:
+        sxfile = path2disp+'sx1_7700Mpc_n6144_nb30_nt16'
+        syfile = path2disp+'sy1_7700Mpc_n6144_nb30_nt16'
+        szfile = path2disp+'sz1_7700Mpc_n6144_nb30_nt16'
+
+    displacements = {}
+    displacements['type'] = 'filelist'
+    displacements['data'] = [sxfile, syfile, szfile]
+
+    return displacements
+
+class FieldSky:
+    '''FieldSky'''
+    def __init__(self, **kwargs):
+
+        import xgfield.defaults as fd
+        self.ID    = kwargs.get(    'ID',fd.ID)
+        self.N     = kwargs.get(     'N',fd.N)
+        self.Lbox  = kwargs.get(  'Lbox',fd.Lbox)
+        self.Nside = kwargs.get( 'Nside',fd.Nside)
+        self.input = kwargs.get( 'input',fd.input)
+        self.nproc = kwargs.get( 'nproc',1)
+        self.rank  = kwargs.get(  'rank',0)
+
+        self.cube  = kwargs.get(  'cube')
+        if self.cube is not None:
+            self.input = 'cube'
+            self.N    = self.cube.N
+            self.Lbox = self.cube.Lbox
+
+        if self.input == 'websky':
+            self.displacements = _get_websky_displacements()
+        elif self.input == 'cube':
+            self.displacements = {}
+            self.displacements['type']  = 'arraylist'
+            self.displacements['data']  = [cube.s1x,cube.s1y,cube.s1z,cube.s2x,cube.s2y,cube.s2z]
+            self.displacements['start'] = (     0,self.N//nproc*rank,         0)
+            self.displacements['end']   = (self.N,self.N//nproc*(rank+1),self.N)
+
+         return
+
+    def generate(self, **kwargs):
+
+        from time import time
+        times={'t0' : time()}
+
+        import os
+        import argparse
+        import logging
+        import xgfield.libfield as lfm
+        import xgfield.defaults as fmd
+        import xgcosmo.cosmology   as cosmo
+        import xgutil.backend      as bk
+
+        log = logging.getLogger("LIGHTCONE")
+
+        grid_nside = self.N # cube shape is parameterized by grid_nside; full resolution for websky is 6144
+        map_nside  = self.Nside
+        L_box      = self.Lbox
+
+        # ------ hardcoded parameters
```

Comment:
@1cosmologist the defaults enter in two places: the [command line interface ](https://github.com/exgalsky/xgfield/blob/cli/xgfield/command_line_interface.py#L16-L21) and in the [initialization of the FieldSky class](https://github.com/exgalsky/xgfield/blob/cli/xgfield/fieldsky.py#L30-L34), since these are the two possible entry points to the `generate` method of `FieldSky`.

The lines here set the values used in `generate` whether they are the default values or were specified on the command line or when initializing the class. The variables in the defaults module are not meant to be changed at runtime, and since the values assigned to these variables can be set by the user at run time, they should remain outside of the defaults module.

Reply:
Ah, yes, I think we should eventually get those into defaults, thanks!